h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/remote/SteamExecutorStarter.java

```diff
@@ -0,0 +1,169 @@
+    private ClusterInfo cluster;
+
+    private final Map<String, Map<String, String>> receivedMessages = new HashMap<>();
+    
+    public SteamExecutorStarter() {
+        instance = this;
+    }
+
+    public RemoteXGBoostExecutor getRemoteExecutor(XGBoostModel model, Frame train, Job<XGBoostModel> job) throws IOException {
+        synchronized (clusterLock) {
+            if (cluster == null) {
+                LOG.info("Starting external cluster for model " + model._key + ".");
+                startCluster(job);
+            } else {
+                LOG.info("External cluster available, starting model " + model._key + " now.");
+            }
+            return makeExecutor(model, train);
+        }
+    }
+
+    private void startCluster(Job<XGBoostModel> job) throws IOException {
+        Map<String, String> startRequest = makeStartRequest();
+        sendMessage(startRequest);
+        Map<String, String> response = waitForResponse(startRequest, job);
+        if (response != null && response.get("status").equals("started")) {
+            String remoteUri = response.get("uri");
+            String userName = response.get("user");
+            String password = response.get("password");
+            cluster = new ClusterInfo(remoteUri, userName, password);
+            LOG.info("External cluster started at " + remoteUri + ".");
+        } else {
+            throw new IllegalStateException("Failed to start external cluster: " + response.get("reason"));
+        }
+    }
+
+    private RemoteXGBoostExecutor makeExecutor(XGBoostModel model, Frame train) {
+        return new RemoteXGBoostExecutor(model, train, cluster.uri, cluster.userName, cluster.password);
+    }
+    
+    private Map<String, String> waitForResponse(Map<String, String> startRequest, Job<XGBoostModel> job) {
+        String responseKey = startRequest.get(ID) + "_response";
+        synchronized (receivedMessages) {
+            while (!receivedMessages.containsKey(responseKey) && !job.stop_requested()) {
+                try {
+                    receivedMessages.wait(10000);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+            return receivedMessages.remove(responseKey);
+        }
+    }
+
+    @Override
+    public void onConnectionStateChange(SteamMessageSender sender) {
+        synchronized (sendingLock) {
+            this.sender = sender;
+        }
+    }
+
+    private synchronized void sendMessage(Map<String, String> message) throws IOException {
+        synchronized (sendingLock) {
+            if (this.sender != null) {
+                sender.sendMessage(message);
+            } else {
+                throw new IOException("Steam communication chanel is not open.");
+            }
+        }
+    }
+
+    @Override
+    public void onMessage(Map<String, String> message) {
+        if (message.get(TYPE).equals("stopXGBoostClusterNotification")) {
```

Comment:
this might be like "stopXGBoostClusterNotification".equals(message.get(TYPE)) to prevent NPE

Reply:
you're right, thats the better way to do it even though type is guaranteed to be non-null by onMessage contract